[Ethan Chua] iP

RCPilot1604 opened the pull request.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

RCPilot1604 committed:
Modify README and greeting page

RCPilot1604 committed:
Modified README and greeting page

RCPilot1604 committed:
Merge remote-tracking branch 'origin/master'

RCPilot1604 committed:
Add level-1 modifications

RCPilot1604 committed:
Add level-2 modifications

Modified Erika bot to do the following:
* Accept text from terminal
* Add text as list items to list
* Print list

RCPilot1604 committed:
Add level-3 modifications

Modified Erika bot to do the following:
* Mark/Unmark list elements

Added Task.java for OOP implementation

RCPilot1604 committed:
Remove stray character

RCPilot1604 committed:
Modify code to comply with coding standards

RCPilot1604 committed:
Implement ToDos, Events and Deadlines using inheritance

RCPilot1604 committed:
Add automated UI testing using I/O redirection

RCPilot1604 committed:
Refactor code to adopt SLAP best practices

RCPilot1604 committed:
Fix bugs in handleMark method

Previously the handleMark method used regex matching to remove
non-digits.

This allowed index 0 to be selected as a valid index although the
indexing of the array starts at 1.

The method was modified to ensure that digits passed were greater
than 0.

wahkia commented on src/main/java/Erika.java:
```diff
@@ -0,0 +1,162 @@
+        int indexOfFrom = line.indexOf("/from ");
+        int indexOfTo = line.indexOf("/to ");
+        if (indexOfFrom != -1 && indexOfTo != -1) {
+            Event newEvent = new Event(line.substring(line.indexOf(" ")+1, indexOfFrom-1),
+                    line.substring(indexOfFrom+6, indexOfTo-1), line.substring(indexOfTo+4));
```
spacing issues

few more similar issues below

RCPilot1604 replied:
Thanks for pointing this out!

wahkia commented on src/main/java/Erika.java:
```diff
@@ -0,0 +1,162 @@
+public class Erika {
+    private static ArrayList<Task> tasks = new ArrayList<>();
+    private static Scanner in = new Scanner(System.in);
+    private static int markIndex;
```
Variable declared follow the camelCase convention, adhering to the Java naming standards. This is good!

wahkia commented on src/main/java/Erika.java:
```diff
@@ -0,0 +1,162 @@
+        }
+    }
+
+    private static void addEvent(String line) {
```
Maybe you could include some Javadoc comments for the functions you created. It would enhance its maintainability and readability of the code.

AndrewNguyen4 replied:
The specific implementation of this method (and other methods specific to their task type, e.g. addDeadline(), addTodo()) should be put in the corresponding class to improve the effect of OOP in your code.

wahkia commented on src/main/java/Erika.java:
```diff
@@ -0,0 +1,162 @@
+        }
+    }
+
+    private static void printMessage(String message) {
```
Nice function for printing!

wahkia reviewed (commented):
Overall, this is a well-written and structured piece of code that adheres to most coding standards!

AndrewNguyen4 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,47 @@
+        taskArraySize--;
+    }
+
+    public String printString() {
+        return "";
+    }
```
- Class java.lang.Object has a default toString() method. You should override this method instead of creating a new method. This applies for all instances of printString() in this program.
- Confusing method name: I would expect printString() to print the string using System.out.println(), not return a string type object.

AndrewNguyen4 commented on src/main/java/Erika.java:
```diff
@@ -0,0 +1,162 @@
+            if (line.equals("bye")) {
+                printGoodbyeMessage();
+                break;
+            } else if (line.equals("list")) {
+                printList();
+            } else if (line.contains("mark ")) {
+                handleMark(line);
+            } else if (line.contains("todo ")) {
+                addTodo(line);
+            } else if (line.contains("deadline ")) {
+                addDeadline(line);
+            } else if (line.contains("event ")) {
+                addEvent(line);
```
Polymorphism could be used here to reduce the need for handling different cases.

RCPilot1604 replied:
Sorry I don't quite understand how said polymorphism will work since I need to deliberately create a new object of the respective subclass (Todo, Deadline, Event) when adding to the task array. Therefore I am not sure how I can leverage polymorphism to call a single (overriden) method for different cases. Thanks!

AndrewNguyen4 commented on src/main/java/Erika.java:
```diff
@@ -0,0 +1,162 @@
+public class Erika {
+    private static ArrayList<Task> tasks = new ArrayList<>();
+    private static Scanner in = new Scanner(System.in);
+    private static int markIndex;
```
The markIndex member of this class can also serve to keep track of the number of tasks in the array, replacing taskArraySize member of classes below. This helps reduce the redundancy in using members.

RCPilot1604 replied:
Hi thanks for your comments! I would like to clarify this comment as markIndex is currently used to store the item the user wishes to mark via the "mark"/"unmark" command. I'm not sure how it can be recycled to also store the size of the task array. Thanks for any clarification!

RCPilot1604 committed:
Modify Erika.java with PR suggestions

RCPilot1604 committed:
Add exception handling

RCPilot1604 committed:
Modify testing files

RCPilot1604 committed:
Add packages and refactor code

RCPilot1604 committed:
Merge branch 'branch-Level-5'

RCPilot1604 committed:
Add more exception cases for empty description

RCPilot1604 committed:
Add new packages

RCPilot1604 committed:
Replace magic numbers with Settings methods

RCPilot1604 committed:
Merge branch 'branch-A-Packages'

RCPilot1604 committed:
Replace handleMark method with indexOperation method

RCPilot1604 committed:
Add deleteEntry method to handle logic for deletion

RCPilot1604 committed:
Rename handleMark method to markEntry

RCPilot1604 committed:
Add deleteEntry method

RCPilot1604 committed:
Remove handleMark method

RCPilot1604 committed:
Add printDeleteMessage method

RCPilot1604 committed:
Modify printing methods for marking/unmarking

RCPilot1604 committed:
Add decrementTaskArraySize method to Task class

RCPilot1604 committed:
Remove destructor of Task superclass

RCPilot1604 committed:
Add FileFormatErrorException

RCPilot1604 committed:
Update .gitignore

RCPilot1604 committed:
Modify Erika.java

RCPilot1604 committed:
Modify Settings.java

RCPilot1604 committed:
Modify Deadline.java

RCPilot1604 committed:
Modify Event.java

RCPilot1604 committed:
Modify Task.java

RCPilot1604 committed:
Modify Todo.java

RCPilot1604 committed:
Add FileSystem.java

RCPilot1604 committed:
Revert "Remove destructor of Task superclass"

This reverts commit 6865cfba0c459667f61cd70ea5c3496d6fdc7a6b.

RCPilot1604 committed:
Modify Erika.java

RCPilot1604 committed:
Merge branch 'branch-Level-6'

RCPilot1604 committed:
Merge branch 'branch-Level-7'

RCPilot1604 committed:
Resolve merge conflicts

RCPilot1604 committed:
Move FileSystem class into package

RCPilot1604 committed:
Add jar and modify .gitignore

RCPilot1604 committed:
Remove dirty bit implementation for immediate updates to FS

RCPilot1604 committed:
Add current progress

Mahesh1772 commented on src/main/java/erika/task/Todo.java:
```diff
@@ -0,0 +1,17 @@
+
+    @Override
+    public String toString() {
+        return String.format("[T][%s] %s", isDone ? "X":" ", description);
```
Consider using the parent class's `toString` method to apply inheritance principle

Mahesh1772 commented on src/main/java/erika/task/Task.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public static void setTaskArraySize(int size) {
+        taskArraySize = size;
```
Is a setter needed for this use case, since you already seem to have a increment and decrement method

Mahesh1772 commented on src/main/java/erika/task/Event.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Override
+    public String generateFileLine() {
+        return String.format("E,%s,%s,%s,%s\n", isDone ? "1" : "0", description, start, end);
```
Consider implementing the `generateFileLine()` method in the `Task` class and use inheritance for better scalability

Mahesh1772 commented on src/main/java/erika/settings/Settings.java:
```diff
@@ -0,0 +1,15 @@
+package erika.settings;
+
+public class Settings {
+    public static final int FROM_LENGTH_OFFSET = 6;
+    public static final int TO_LENGTH_OFFSET = 4;
+    public static final int BY_LENGTH_OFFSET = 4;
+    public static final int SPACE_OFFSET = 1;
+    public static final int BY_REAR_OFFSET = 1;
+    public static final int FROM_REAR_OFFSET = 1;
+    public static final int TO_REAR_OFFSET = 1;
+
+    public static final String DIRECTORY = "data";
+    public static final String FILENAME = DIRECTORY + "/tasks.txt";
+    public static final String SEPARATOR = ",";
+}
```
Good job defining constants to a file

Mahesh1772 commented on src/main/java/erika/filesystem/FileSystem.java:
```diff
@@ -0,0 +1,106 @@
+                throw new FileFormatErrorException("Error on line " + lineNumber + ": invalid Todo line!");
+            }
+            Todo todo = new Todo(description);
+            todo.setMark(tokens[1].equals("1"));
+            tasks.add(todo);
+            break;
+
+        case "D":
+            if (tokens.length != 4) {
+                throw new FileFormatErrorException("Error on line " + lineNumber + ": invalid Deadline line!");
+            }
+            Deadline deadline = new Deadline(description, tokens[3]);
+            deadline.setMark(Boolean.parseBoolean(tokens[1]));
+            tasks.add(deadline);
+            break;
+        case "E":
+            if (tokens.length != 5) {
+                throw new FileFormatErrorException("Error on line " + lineNumber + ": invalid Event line!");
+            }
+            Event event = new Event(description, tokens[3], tokens[4]);
+            event.setMark(Boolean.parseBoolean(tokens[1]));
```
Consider having a `setMark` method in the `Task` class as to reduce duplicate lines of code in different subclasses

Mahesh1772 commented on src/main/java/erika/Erika.java:
```diff
@@ -0,0 +1,293 @@
+    }
+
+    private static void printMessage(String message) {
+        System.out.println("\t____________________________________________________________");
```
Consider making this line into a constant string, for easy reproducibility and keeping visuals consistent

Mahesh1772 commented on src/main/java/erika/Erika.java:
```diff
@@ -0,0 +1,293 @@
+        try {
+            fileSystem.appendTaskToFile(newTodo);
+        } catch (IOException e) {
+            printMessage("Unable to append Todo to FS. IO Error occurred");
```
Consider adding these error messages into the constants file as well

Mahesh1772 commented on src/main/java/erika/Erika.java:
```diff
@@ -0,0 +1,293 @@
+
+    private static boolean mainLoop() {
+        try{
+            String line = collectUserInput();
+            if (line.equals("bye")) {
+                printGoodbyeMessage();
+                return true;
+            } else if (line.equals("list")) {
+                printList();
+            } else if (line.contains("mark")) {
+                indexOperation(line, false);
+            } else if (line.contains("todo")) {
+                addTodo(line);
+            } else if (line.contains("deadline")) {
+                addDeadline(line);
+            } else if (line.contains("event")) {
+                addEvent(line);
+            } else if (line.contains("delete")) {
+                indexOperation(line, true);
+            } else {
+                throw new UnknownCommandException();
+            }
```
Consider using switch case for clearer code organization and because the values expected are predetermined.

Mahesh1772 reviewed (commented):
Overall, this implementation is solid, with a clear understanding of task management logic, but there are areas for improvement. Consider adding **JavaDocs** in future commits. Keep up the good work!

RCPilot1604 committed:
Add AddCommand classes

RCPilot1604 committed:
Modify main loop to support OOP implementation

RCPilot1604 committed:
Add Command classes to handle command execution

RCPilot1604 committed:
Add Console class

RCPilot1604 committed:
Add Exception classes

RCPilot1604 committed:
Add FileSystem class

RCPilot1604 committed:
Add Parser class

RCPilot1604 committed:
Modify Task class

Removed the static attribute taskArraySize (replaced in TaskList)

RCPilot1604 committed:
Add TaskList class

RCPilot1604 committed:
Add Ui class

RCPilot1604 committed:
Add OutOfBoundsException to replace default exception

This is for more consistent exception handling in the main code

RCPilot1604 committed:
Modify AddDeadlineCommand to support /by LocalDate(Time)

RCPilot1604 committed:
Add DateTimeFormatter settings for input and output

RCPilot1604 committed:
Modify toString and generateFileLine methods to output dates

RCPilot1604 committed:
Move command processing functions to their respective classes

RCPilot1604 committed:
Remove command processing code from Parser

RCPilot1604 committed:
Remove command processing code from TaskList

RCPilot1604 committed:
Add FindCommand class

RCPilot1604 committed:
Javadoc for Command superclass

RCPilot1604 committed:
Add JavaDoc comments to Command

RCPilot1604 committed:
Add JavaDoc comments to DeleteCommand

RCPilot1604 committed:
Add JavaDoc comments to ExitCommand

RCPilot1604 committed:
Add JavaDoc comments to ListCommand

RCPilot1604 committed:
Add JavaDoc comments to MarkCommand

RCPilot1604 committed:
Add JavaDoc comments to UnmarkCommand

RCPilot1604 committed:
Add JavaDoc comments to AddCommand

RCPilot1604 committed:
Add JavaDoc comments to AddDeadlineCommand

RCPilot1604 committed:
Add JavaDoc comments to AddEventCommand

RCPilot1604 committed:
Add JavaDoc comments to AddTodoCommand

RCPilot1604 committed:
Add JavaDoc comments to Console

RCPilot1604 committed:
Add JavaDoc comments to EmptyDescriptionException

RCPilot1604 committed:
Add JavaDoc comments to EmptyListException

RCPilot1604 committed:
Add JavaDoc comments to ErikaException

RCPilot1604 committed:
Add JavaDoc comments to FileFormatException

RCPilot1604 committed:
Add JavaDoc comments to FormatErrorException

RCPilot1604 committed:
Add JavaDoc comments to OutOfBoundsException

RCPilot1604 committed:
Add JavaDoc comments to UnknownCommandException

RCPilot1604 committed:
Add JavaDoc comments to FileSystem

RCPilot1604 committed:
Add JavaDoc comments to Settings

RCPilot1604 committed:
Add JavaDoc comments to Deadline

RCPilot1604 committed:
Add JavaDoc comments to Event

RCPilot1604 committed:
Add JavaDoc comments to Todo

RCPilot1604 committed:
Add JavaDoc comments to TaskList

RCPilot1604 committed:
Merge pull request #1 from RCPilot1604/branch-A-MoreOOP

Branch-A-MoreOOP

RCPilot1604 committed:
Merge pull request #2 from RCPilot1604/branch-Level-8

Branch-Level-8

RCPilot1604 committed:
Merge branch 'master' into branch-Level-9

# Conflicts:
#	src/main/java/erika/command/addcommand/AddDeadlineCommand.java

RCPilot1604 committed:
Merge pull request #3 from RCPilot1604/branch-Level-9

Branch-Level-9

RCPilot1604 committed:
Merge branch 'master' into branch-A-JavaDoc

# Conflicts:
#	src/main/java/erika/command/DeleteCommand.java
#	src/main/java/erika/command/ListCommand.java
#	src/main/java/erika/command/MarkCommand.java
#	src/main/java/erika/command/UnmarkCommand.java
#	src/main/java/erika/command/addcommand/AddCommand.java
#	src/main/java/erika/command/addcommand/AddDeadlineCommand.java
#	src/main/java/erika/settings/Settings.java
#	src/main/java/erika/task/Deadline.java
#	src/main/java/erika/task/Event.java
#	src/main/java/erika/tasklist/TaskList.java

RCPilot1604 committed:
Merge pull request #4 from RCPilot1604/branch-A-JavaDoc

Branch-A-JavaDoc

RCPilot1604 committed:
Modify Settings to enable more robust error-checking of commands

RCPilot1604 committed:
Update UI redirection tests

RCPilot1604 committed:
Fix bugs

RCPilot1604 committed:
Add user guide introduction

RCPilot1604 committed:
Add deadline section to UG

RCPilot1604 committed:
Add event section to UG

RCPilot1604 committed:
Add delete section to UG

RCPilot1604 committed:
Add mark and unmark sections to UG

RCPilot1604 committed:
Add listing section to UG

RCPilot1604 committed:
Add saving section to UG

RCPilot1604 committed:
Merge branch 'branch-UG'

RCPilot1604 committed:
Add support for delete all command

Previously accidentally removed by version-control

RCPilot1604 committed:
Modify Parser to fix bug in commands

Previously, the commands were matched with keywords sequentially.
This created an edge case when command lower on the hierarchy included
a keyword from a command higher on the hierarchy.
For instance, "find todo" would be interpreted as a todo command due
to "todo" being matched first.

To fix this, the condition for matching "COMMAND KEYWORD" is modified
to check that said keyword is at the first index of the line.

RCPilot1604 committed:
Modify TaskList to support delete all command

RCPilot1604 committed:
Modify Javadoc comments to conform to coding standard

RCPilot1604 committed:
Modify Javadoc comments to comply with coding standards

RCPilot1604 committed:
Update Javadoc comments

RCPilot1604 committed:
Update .gitignore

RCPilot1604 committed:
Update .gitignore

RCPilot1604 committed:
Add NoSuchElementException handling

RCPilot1604 committed:
Modify Deadline class to standardize with Event class

RCPilot1604 committed:
Modify Parser class to enable handling of missing spaces

RCPilot1604 committed:
Modify Settings class to handle different dateTime formats

RCPilot1604 committed:
Modify deadline to standardize with Event

RCPilot1604 committed:
Modify Event class to utilize Settings parser

RCPilot1604 committed:
Modify IO redirection tests

RCPilot1604 committed:
Modify Erika to handle NumberFormatException

RCPilot1604 committed:
Modify logic to distinguish between exceptions

Distinguish between FormatErrorException and EmptyDescriptionException

RCPilot1604 committed:
Add more date formats

RCPilot1604 committed:
Update IO testing

RCPilot1604 committed:
Merge pull request #5 from RCPilot1604/branch-test

Merge branch-test into master

RCPilot1604 committed:
Merge remote-tracking branch 'origin/master'

RCPilot1604 committed:
Prune accepted DateTime formats for clarity

RCPilot1604 committed:
Update README.md

RCPilot1604 committed:
Fix bugs in unmark command

RCPilot1604 committed:
Update frontend .README.md

RCPilot1604 committed:
Update README.md 

Add link to user guide

RCPilot1604 committed:
Fix formatting issues

RCPilot1604 committed:
Modify IO tests

RCPilot1604 committed:
Update Erika to handle empty parser lines during IO testing

RCPilot1604 committed:
Update .gitignore

RCPilot1604 committed:
Merge branch 'add-gradle-support' of https://github.com/RCPilot1604/ip

RCPilot1604 committed:
Update NumberFormatException error message

RCPilot1604 committed:
Revert "Merge branch 'add-gradle-support' of https://github.com/RCPilot1604/ip"

This reverts commit fae9aab7fc7558b9b750290861d0eb3bbafd9962, reversing
changes made to 958f7fbcaa228624c07317c3c8d762ffb741cda8.

RCPilot1604 committed:
Modify NumberFormatException error message

RCPilot1604 committed:
Modify code to comply with coding standard

RCPilot1604 committed:
Modify formatting to comply with coding standard

RCPilot1604 committed:
Modify format to comply with coding standard

RCPilot1604 committed:
Remove input.txt from tracking list

RCPilot1604 committed:
Modify README.md

RCPilot1604 committed:
Update README.md

RCPilot1604 committed:
Update README.md
